KAFKA-6958: Allow to define custom processor names with KStreams DSL

fhussonnois opened the pull request:
This is a WIP for the KIP :  https://cwiki.apache.org/confluence/display/KAFKA/KIP-307%3A+Allow+to+define+custom+processor+names+with+KStreams+DSL

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/ForeachAction.java:
```diff
@@ -38,6 +38,29 @@
+    static <K, V> ForeachAction<K, V> named(final String name, final ForeachAction<K, V> action) {
+        return new ForeachAction<K, V>() {
+            @Override
+            public void apply(K key, V value) {
```
nit: `final` on parameters here and others below

bbejeck reviewed (commented):
Thanks for the contribution @fhussonnois!  

I understand this is still a WIP so I just took a quick pass over the PR and it looks good so far.  We should also have some unit tests confirming the names are set as expected (forgive me for stating the obvious!)

guozhangwang commented:
@fhussonnois please ping me and @bbejeck whenever you think that the PR is ready for final reviews.

mjsax commented:
@guozhangwang What do you mean by "final review" -- the KIP is not voted yet, but still in DISCUSS stage.

guozhangwang commented:
> @guozhangwang What do you mean by "final review" -- the KIP is not voted yet, but still in DISCUSS stage.

For such KIPs it is better to do PR reviews and KIP discussions in parallel as some details may not be seen on email discussion, so I'd like to take a look at the PR as early as possible as we still in DISCUSS stage. But even with the current proposal, it seems the PR is not complete yet, so I left the above comment :)

mjsax commented:
That makes sense. Was just irritated by "final" :)

fhussonnois commented:
@guozhangwang @bbejeck thanks for your review on this PR. To be honest I'm not sure to have covered everything.

mjsax commented:
Build failed with compile errors.

fhussonnois commented:
Compile issues and checkstyles have been fixed.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java:
```diff
@@ -0,0 +1,127 @@
+     * {@inheritDoc}
+     *
+     * @throws IllegalStateException if internal name is null.
+     * @return  a string name.
```
This seems too restrict to me: for users upgrading to the new version, the passed in class which extends `Named` interface would have empty named string, and hence the constructor at line 46 above will doom to throw at `.name()` function call.

fhussonnois replied:
An exception will be thrown if you do something like : new NamedInternal(new NamedInternal(null)), but why doing this ? The method NamedInternal#name() should only be called internally when we do expect a name. I think that throwing an IllegalStateException here allow to protect against a bad used of the NamedInternal class. 

Users should never have this IllegalStateException (except if they do use the NamedInternal class which is not a public API).

guozhangwang replied:
I was primarily thinking about the code simplicity. More specifically:

1) Examples like https://github.com/apache/kafka/pull/5909/files/1a61d17d40f9f315689c978bba3dbfd50e3f22ef#diff-0550fd2580d044bc4c2a8fa6cfe55f6cR91 where we have to make sure the NamedInternal's name is not null in the callers, whereas previously we can just construct it inside the `build()` call. That means each caller need to get `newProcessorName()` by itself.

2. Similarly, we need the validation call on each of the caller as well. Whereas I was suggesting we can just make it an internal check which is deferred after the constructor call.

WDYT?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:
```diff
@@ -225,8 +228,14 @@
+        final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materializedInternal =
+                new MaterializedInternal<>(materialized);
+
+        final NamedInternal namedInternal = new NamedInternal(consumedInternal.processorName());
```
Follow-up on my other comment on `NamedInternal`: if we can allow the passed in `Named` object in the constructor to not contain a processor name, then we can simply do 

```
namedInternal = new NamedInternal(consumed);
```

here. Ditto elsewhere.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:
```diff
@@ -399,8 +418,10 @@
         final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materializedInternal = new MaterializedInternal<>(materialized);
-        materializedInternal.generateStoreNameIfNeeded(internalStreamsBuilder, topic + "-");
 
+        final NamedInternal namedInternal = NamedInternal.with(consumedInternal.processorName());
```
This call trace is different with others above: is this intentional?

Personally I think either constructor or `NamedInternal.with()` is fine, just calling to be consistent here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -406,6 +406,23 @@
+     * @return multiple distinct substreams of this {@code KStream}
+     */
+    @SuppressWarnings("unchecked")
+    KStream<K, V>[] branch(final Named processorName, final Predicate<? super K, ? super V>... predicates);
```
nit: change parameter name to `named` for consistency. Ditto elsewhere and in `KTable` as well.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -293,18 +293,31 @@ public void print(final Printed<K, V> printed) {
         final String[] childNames = new String[predicates.length];
         for (int i = 0; i < predicates.length; i++) {
-            childNames[i] = builder.newProcessorName(BRANCHCHILD_NAME);
+            childNames[i] = new NamedInternal(predicates[i]).orElseGenerateWithPrefix(builder, BRANCHCHILD_NAME);
```
We should leverage on the predicate's named field here instead of still using the branch processor name. More specifically, we create one processor node for the branch operator, and n children pass-through node. Only the former should be leverage on the passed in name, while for the children name we should base on the corresponding predicate name.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -293,18 +293,31 @@ public void print(final Printed<K, V> printed) {
+
+    @SuppressWarnings("unchecked")
+    private KStream<K, V>[] doBranch(final NamedInternal processorName,
+                                     final Predicate<? super K, ? super V>... predicates) {
```
Why no `NamedInternal.validate` here? Ditto on some other functions.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -373,7 +394,7 @@ public void foreach(final ForeachAction<? super K, ? super V> action) {
     public KStream<K, V> peek(final ForeachAction<? super K, ? super V> action) {
         Objects.requireNonNull(action, "action can't be null");
-        final String name = builder.newProcessorName(PEEK_NAME);
+        final String name = new NamedInternal(action).orElseGenerateWithPrefix(builder, PEEK_NAME);
```
Why no `NamedInternal.validate` here? Ditto on some other functions.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -469,24 +502,43 @@ private void to(final TopicNameExtractor<K, V> topicExtractor, final ProducedInt
+        NamedInternal.validate(name);
         final StatefulProcessorNode<? super K, ? super V> transformNode = new StatefulProcessorNode<>(
-            name,
+                name,
```
nit: this seems not necessary.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -565,13 +628,18 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
         KStreamImpl<K, V> joinThis = this;
         KStreamImpl<K, VO> joinOther = (KStreamImpl<K, VO>) other;
 
+        final NamedInternal joinedName = new NamedInternal(joined);
```
validation missed?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -883,12 +953,26 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
-
+            final NamedInternal namedInternal = new NamedInternal(joiner);
+
+            final String thisWindowStreamName =  namedInternal.suffixWithOrElseGet(
```
I'd suggest we use the suffix of those nodes as `-this/other-windowed` and `-this/other-join` and `-merge` for simplicity.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -117,7 +120,7 @@ public String queryableStoreName() {
                                   final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materializedInternal,
                                   final boolean filterNot) {
-        final String name = builder.newProcessorName(FILTER_NAME);
+        final String name = new NamedInternal(predicate).orElseGenerateWithPrefix(builder, FILTER_NAME);
```
validation missed?

guozhangwang replied:
Actually I'd suggest we make validation private and called internally (see my other comment on NamedInternal).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java:
```diff
@@ -0,0 +1,127 @@
+     *
+     * @throws InvalidProcessorException if the processor name is invalid.
+     */
+    public static void validate(final String processorName) {
```
Making it a public function would require us to call `validate` on all the places scattered across KTableImpl / KStreamImpl etc classes. Instead I'd suggest we just make it a private function and call it in

1) constructor which takes the String or Named.
2) static function that creates a Named from String.

By doing this we do not need to call this elsewhere, this is because for library created names internally, we are sure that they are always valid so we do not need to validate them.

guozhangwang replied:
BTW: In that case, we would allow the internal name to be null.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -264,34 +272,53 @@ public String queryableStoreName() {
         Objects.requireNonNull(materialized, "materialized can't be null");
         final MaterializedInternal<K, VR, KeyValueStore<Bytes, byte[]>> materializedInternal = new MaterializedInternal<>(materialized);
-        materializedInternal.generateStoreNameIfNeeded(builder, TRANSFORMVALUES_NAME);
+        final NamedInternal namedInternal = new NamedInternal(processorName.name());
```
ditto here and elsewhere: we can just call `new NamedInternal(processorName)` (btw I suggest rename it to `renamed`, see my other comment) if we do not restrict its internal name to be not null.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -358,11 +385,12 @@ public String queryableStoreName() {
-        final String storeName =
-            suppressedInternal.name() != null ? suppressedInternal.name() + "-store" : builder.newStoreName(SUPPRESS_NAME);
+        final String name = suppressedName.orElseGenerateWithPrefix(builder, SUPPRESS_NAME);
+        NamedInternal.validate(name);
```
As long as `suppressedName` is valid, its derived storeName should always be valid, so we only need to valid once, and better be internal at the constructor (see my other comment).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/ProducedInternal.java:
```diff
@@ -36,4 +36,8 @@ public ProducedInternal(final Produced<K, V> produced) {
         return partitioner;
     }
+
+    public String processorName() {
```
This is a meta comment: I'd suggest let other XXXInternal to extends XXX and NamedInternal so that this function is consolidated on the base class as `public String name()` which is allowed to return null.

fhussonnois replied:
I think what you mean is that Produced/Consumed classes should extend Named ? Because XXXInternal classes already extend XXX (no multiple inheritance)

guozhangwang replied:
Ah yes, you are right. Hmm then that means we need to move this function to `Named` so that:

Named (contains `processorName()`) -> Consumed / etc -> ConsumedInternal / etc (shares `processorName()`

But this is a bit awkward since all other `Consumed / etc` public classes does not have getters only setters.

So how about making this function package-private while moving to Named, so that it will not be exposed as public APIs but still be able to accessed internally on XXXInternal callers?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/Consumed.java:
```diff
@@ -129,7 +137,19 @@ protected Consumed(final Consumed<K, V> consumed) {
+     * @param <V>         value type
+     * @return a new instance of {@link Consumed}
+     */
+    public static <K, V> Consumed<K, V> with(final String processorName) {
```
Personally I like to rename it (and others on the other classes) to `as`. I know this is a subjective preference and hence would like to hear more thoughts? @fhussonnois @mjsax @vvcephei @bbejeck

fhussonnois replied:
if we rename to `as` we should also change the Suppressed interface to rename `withName` to `as` for consistency ?

guozhangwang replied:
Good call. Actually there are also `Grouped` and `Joined` which are already released as part of 2.1.

So how about renaming it to `withName` instead of `with` for consistency with Grouped, Joined, and Suppresed? cc @vvcephei @bbejeck

fhussonnois replied:
Actually : 
-  `Grouped` class has a static method `as` and instance method `withName`
- `Joined` class has a static method  `named` and instance method `withName`
- `Suppressed` class has no static method to defined name and instance method `withName`

So it's OK to rename the static methods `with` to `as` but we should change the `Joined` class too.

guozhangwang replied:
I think this is a good point. It is a regret that 2.1 is already released with `Joined.named` but personally I think it still makes more sense to predicate and replace it with `Joined.as`, while @fhussonnois get all others to be consistent with `as`. @bbejeck wdyt?

guozhangwang replied:
Hey @fhussonnois @vvcephei @bbejeck  I've made one more pass over all the control classes we have so far, and here is a list along with the current state method name:

0. Named: `as` (new, Aggregator / Mapper / Predicate / Foreach / Joiner / and SuppressedInternal extends it)
1. Produced / Consumed: `with` (new)
2. Printed: `withName` (new)
3. Grouped: `as`, BUT this is used for repartition topic names only
4. Joined: `named`, BUT this is used for repartition topic names only
5. Suppressed: `withName`, BUT this is used for both processor name and store name
6. Materialized: `as`, BUT this is used for state store names only

So as I read it, the semantics of 3/4/6 are not correlated to 0/1/2, as they are used for naming the processors only, not repartition topics / state stores.

In my other comment I was wondering if we could have 1/2 extending 0 as well, because the operators leveraging them do not have a `Aggregator / etc` lambda that extends Named so we are sort of using those three as a "shortcut" to still allow users to override names. For suppression operator, it is the similar rationale (i.e. no transformation and hence no lambda functions, just leverage on the control object). Following this, I think 5 can also extend 0 as well (in your current impl only `SuppressedInternal` extends `NamedInternal`, I'd suggest we let `Named` -> `Suppressed` -> `SuppresedInternal` and moving `Named.processorName()` as package private, same as 1/2). 

-------------------------------------

So I'd suggest let Produced / Consumed / Printed / Suppressed to extend Named. Move `processorName` to Named as package private so that all sub-classes can use. Keep `Named.as` function name across these functions as well because `Named.withName` or `Named.named` sounds quite awkward.

-------------------------------------

For 6, a.k.a. `Materialized.as` is really for a different scope and orthogonal to this discussion, so I'd suggest we keep it as-is. 

-------------------------------------

For 3/4, this is a bit tricky since now we are effectively providing multiple ways to allow users to set repartition topics. To be more concrete, for example:

a) `stream1.join(stream2, named("name1", /* join logic */))` -> use `name1` for processor, and also `name1-repartition` as the repartition topic.

b) `stream1.join(stream2, /* join logic*/, Joined.named("name2"))` -> use `name2` for repartition topic, internal name for processor.

c) `stream1.join(stream2, named("name1", /* join logic */), Joined.named("name2"))` -> use `name1` for processor name, and `name2-repartition` for repartition name.

This sounds to me a bit too confusing to users, and hence makes me wondering if we should just keep one of the two set of APIs. WDYT?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/Named.java:
```diff
@@ -0,0 +1,42 @@
+     * @param name  the name to be used.
+     * @return a new {@link Named} instance .
+     */
+    static Named with(final String name) {
```
Similar to the other comment: I'd suggest rename the function to `as`.

guozhangwang reviewed (commented):
Thanks for the PR! I've made a pass over it. So far we do not allow users to provide processor names in:

1. KTable#toStream()
2. KGroupedTable / KGroupedStream#count()

I think this is fine but like to hear other's thoughts.

Also another meta comment is that we still lack the corresponding doc changes on streams dsl under developer guide, and most importantly the upgrade guide (I think with this change we need users to recompile their code as it is not binary compatible, and we need to make that clear).

vvcephei commented:
Hi @fhussonnois ,

Just a heads-up, I sent a message to the discussion thread on the mailing list with some high-level questions about the approach.

Thanks for the PR and KIP! This will be a really nice feature.
-John

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Grouped.java:
```diff
@@ -124,13 +124,19 @@ protected Grouped(final Grouped<K, V> grouped) {
      * @return a new {@link Grouped} instance configured with the name
      * */
+    @Override
     public Grouped<K, V> withName(final String name) {
```
JavaDocs (also other methods) needs to be updated: the name is not only used for the repartition topic but also as Processor name now.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -99,7 +170,14 @@
+     * <p>
+     *
      * @param mapper a {@link KeyValueMapper} that computes a new key for each record
+     * @param named  the name to be used for the selectKey node
```
nit: `node` -> `processor` ? (similar in other methods)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -406,6 +803,23 @@
+     * A record will be dropped if none of the predicates evaluate to true.
+     * This is a stateless record-by-record operation.
+     *
+     * @param named the name to be used for processor
```
nit `for [the] processor` (similar in some other methods)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:
```diff
@@ -530,7 +1033,7 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      *                 // punctuate each 1000ms; can access this.state
      *                 // can emit as many new KeyValue pairs as required via this.context#forward()
-     *                 context.schedule(Duration.ofSeconds(1), PunctuationType.WALL_CLOCK_TIME, new Punctuator(..));
+     *                 context.schedule(1000, PunctuationType.WALL_CLOCK_TIME, new Punctuator(..));
```
Why this change? The new API accepts `Duration` (old API that accepts a long is deprecated).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Named.java:
```diff
@@ -0,0 +1,94 @@
+
+public class Named implements NamedOperation<Named> {
+
+    private static final int MAX_NAME_LENGTH = 249;
```
Where does this number come from?

fhussonnois replied:
This number come from the class [Topic](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/internals/Topic.java)

mjsax replied:
Well. Then we should make it public in `Topic` and reuse here -- otherwise, we have code duplication and code might diverge.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Named.java:
```diff
@@ -0,0 +1,94 @@
+     * @throws InvalidProcessorException if the processor name is invalid.
+     */
+    private static void validate(final String processorName) {
+        if (processorName.isEmpty())
```
nit: add `{ }` to all blocks.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Named.java:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @throws InvalidProcessorException if the processor name is invalid.
+     */
+    private static void validate(final String processorName) {
```
I am wondering if there is `validate` method in a common package already that we could reuse instead?

fhussonnois replied:
As above, this method actually comes from https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/internals/Topic.java#L36. 
But I was faced with two concerns

1) Should we not try to minimize dependencies between streams API and internal clients API ?

2) The method validate throws an InvalidTopicException. Even if I can catch this exception I will loose the error message details if we want to hide the relationship with "topic".

WDYT?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/Named.java:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * Valid characters for Kafka topics are the ASCII alphanumerics, '.', '_', and '-'
+     */
+    private static boolean containsValidPattern(final String topic) {
```
Similar as above. Can we reuse some existing code?

fhussonnois replied:
As above, the method comes from https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/internals/Topic.java#L77.

If we are OK to create a dependency between this two classes we jut make this method public.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -128,28 +130,36 @@
-                                 repartitionRequired,
-                                 filterProcessorNode,
-                                 builder);
+                keySerde,
```
nit: move `name` into it's own line to get equal indention for all parameters.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -229,20 +259,19 @@
 
         // value serde cannot be preserved
         return new KStreamImpl<>(name,
```
as above.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -505,12 +611,22 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
+
+    @Override
+    public void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
+                 final Named processorName,
```
nit: `processorName` seems to be an inconsistent parameter name compare to other methods adding a `Named` parameter.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -505,12 +611,22 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
-            new ProcessorParameters<>(processorSupplier, name),
-            stateStoreNames,
-            repartitionRequired
+                name,
```
nit: indention should be 4 spaces, not 8.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/MaterializedInternal.java:
```diff
@@ -33,13 +34,19 @@ public MaterializedInternal(final Materialized<K, V, S> materialized) {
+    public void setStoreNameIfNeeded(final Supplier<String> storeNameSupplier) {
         if (storeName() == null) {
-            storeName = nameProvider.newStoreName(generatedStorePrefix);
+            this.storeName = storeNameSupplier.get();
```
nit: remove `this` (we use `this` only if required)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedKStreamImpl.java:
```diff
@@ -122,33 +148,51 @@
-            materializedInternal.isQueryable(),
-            materializedInternal.keySerde() != null ? new WindowedSerdes.SessionWindowedSerde<>(materializedInternal.keySerde()) : null,
-            materializedInternal.valueSerde());
+                new NamedInternal(reduceName),
```
nit: indention should be 4 spaces

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowedKStreamImpl.java:
```diff
@@ -115,23 +145,39 @@
-            materializedInternal.isQueryable(),
-            materializedInternal.keySerde() != null ? new FullTimeWindowedSerde<>(materializedInternal.keySerde(), windows.size()) : null,
-            materializedInternal.valueSerde());
+                new NamedInternal(aggregateName),
```
nit: indention

mjsax reviewed (commented):
Thanks for the WIP PR. This helps a lot!

Did a quick high-level pass over the PR to get an impression of the change. Will follow up on the mailing list.

mjsax commented:
@fhussonnois Would it be possible to split this into multiple PRs (like 5 to 10)? It's huge and thus very hard to review, what will slow down the review time significantly.

fhussonnois committed:
KAFKA-6958: Add new NamedOperation interface to enforce consistency in naming operations

Sub-task required to allow to define custom processor names with KStreams DSL(KIP-307) :
  - add new public interface NamedOperation
  - deprecate methods Joined.as() and Joined.name()
  - update Suppredded interface to extend NamedOperation

fhussonnois committed:
KAFKA-6958: Allow to name operation using parameter classes

Sub-task required to allow to define custom processor names with KStreams DSL(KIP-307) :
 - update existing classes Consumed/Grouped/Printed/Produced to implement NamedOperation
 - introduce new public class Named

fhussonnois committed:
KAFKA-6958: Overload KStream methods to allow to name operation name using the new Named class

Sub-task required to allow to define custom processor names with KStreams DSL(KIP-307) :
 - overload methods for stateless operations to accept a Named parameter (filter, filterNot, map, mapValues, foreach, peek, branch, transform, transformValue, flatTransform)
 - overload process method to accept a Named parameter
 - overload join/leftJoin/outerJoin methods

fhussonnois committed:
KAFKA-6958: Overload KTable methods to allow to name operation name using the new Named class

Sub-task required to allow to define custom processor names with KStreams DSL(KIP-307) :

fhussonnois committed:
KAFKA-6958: Overload methods for group and windowed stream to allow to name operation name using the new Named class

Sub-task required to allow to define custom processor names with KStreams DSL(KIP-307) :
 - overload method for KGroupedStream to accept a Named parameter
 - overload method for KGroupedTable to accept a Named parameter
 - overload method for TimeWindowedKStream to accept a Named parameter
 - overload method for SessionWindowedKStream to accept a Named parameter

mjsax commented:
Closing this because it seems outdate. The work was broken into multiple smaller PRs.
